parachains/pallets/bridge-assets-transfer/src/lib.rs

```diff
@@ -0,0 +1,888 @@
+			let allowed_target_location_network = bridge_config
+				.allowed_target_location
+				.interior()
+				.global_consensus()
+				.map_err(|_| Error::<T>::InvalidConfiguration)?;
+			ensure!(
+				bridged_network == allowed_target_location_network,
+				Error::<T>::InvalidConfiguration
+			);
+
+			Bridges::<T>::insert(bridged_network, bridge_config);
+			Self::deposit_event(Event::BridgeAdded);
+			Ok(())
+		}
+
+		/// Remove bridge configuration for specified `bridged_network`.
+		///
+		/// Parameters:
+		///
+		/// * `bridged_network`: Network where we want to remove
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::remove_bridge_config())]
+		pub fn remove_bridge_config(
+			origin: OriginFor<T>,
+			bridged_network: NetworkId,
+		) -> DispatchResult {
+			let _ = T::AdminOrigin::ensure_origin(origin)?;
+			ensure!(Bridges::<T>::contains_key(bridged_network), Error::<T>::InvalidConfiguration);
+
+			Bridges::<T>::remove(bridged_network);
+			Self::deposit_event(Event::BridgeRemoved);
+			Ok(())
+		}
+
+		/// Updates bridge configuration for specified `bridged_network`.
+		///
+		/// Parameters:
+		///
+		/// * `bridged_network`: Network where we want to remove
+		/// * `fee`: New fee to update
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::update_bridge_config())]
+		pub fn update_bridge_config(
+			origin: OriginFor<T>,
+			bridged_network: NetworkId,
+			fee: Option<MultiAsset>,
+		) -> DispatchResult {
+			let _ = T::AdminOrigin::ensure_origin(origin)?;
+			ensure!(Bridges::<T>::contains_key(bridged_network), Error::<T>::InvalidConfiguration);
+
+			Bridges::<T>::try_mutate_exists(bridged_network, |bridge_config| {
+				let deposit = bridge_config.as_mut().ok_or(Error::<T>::InvalidConfiguration)?;
+				deposit.fee = fee;
+				Self::deposit_event(Event::BridgeUpdated);
+				Ok(())
+			})
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Validates destination and check if we support bridging to this remote global consensus
+		///
+		/// Returns: correct remote location, where we should be able to bridge
+		pub(crate) fn ensure_remote_destination(
+			remote_destination: VersionedMultiLocation,
+		) -> Result<(BridgeConfig, MultiLocation), Error<T>> {
+			match remote_destination {
+				VersionedMultiLocation::V3(remote_location) => {
+					ensure!(
+						remote_location.parent_count() == 2,
+						Error::<T>::UnsupportedDestination
+					);
+					let local_network = T::UniversalLocation::get()
+						.global_consensus()
+						.map_err(|_| Error::<T>::InvalidConfiguration)?;
+					let remote_network = remote_location
+						.interior()
+						.global_consensus()
+						.map_err(|_| Error::<T>::UnsupportedDestination)?;
+					ensure!(local_network != remote_network, Error::<T>::UnsupportedDestination);
+					match Bridges::<T>::get(remote_network) {
+						Some(bridge_config) => {
+							ensure!(
+								remote_location.starts_with(&bridge_config.allowed_target_location),
```

Comment:
This case is tricky for Ethereum, which doesn't have hierarchical consensus locations. In that network, its perfectly valid for the remote destination not to start with the allowed target location.

For example:
```
// recipient and executor of xcm message
allowed_target_location: /GlobalConsensus(Ethereum)/AccountKey20(AssetManagerContract) 
// final destination for funds
remote_destination: /GlobalConsensus(Ethereum)/AccountKey20(Alice)
```

Reply:
@vgeddes
I am trying to have Ethereum  scenario in my mind,
result of this call `transfer_asset_via_bridge` is `ExportMessage(ReserveAssetDeposited)` to be send to the BridgeHub,
where we already have `MessageExporter` https://github.com/paritytech/cumulus/blob/bridge-hub-rococo-wococo/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L245
so for Ethereum we just need to add new one:
```
type MessageExporter = (BridgeHubRococoOrBridgeHubWococoSwitchExporter, EthereumExporter);
```
I assume that EthereumExporter should just "transform/process" `XCM(ReserveAssetDeposited)` and feed some "eth-pallet" from which some eth-relayer takes it and call eth smart-contract, right?



so back to the `ExportMessage` generation: 

---
lets says for Statemine -> Statemint transfer for KSM, this will generate:

```
allowed_target_location: MultiLocation::new(2, X3(GlobalConsensus(Polkadot), Parachain(1000))


transfer_asset_via_bridge(
    // KSM asset from Statemine point of view
    MultiAsset {
        fun: Fungible(1000.into()),
        id: Concrete(MultiLocation { parents: 1, interior: Junctions::Here }),
    },
    // remote destination
    MultiLocation::new(2, X3(GlobalConsensus(Polkadot), Parachain(1000), consensus_account(Polkadot, 2)))
)
```

```
Xcm([
    UnpaidExecution { weight_limit: Unlimited, check_origin: None },
    // KSM relay location reanchored how Statemint will see it
    ReserveAssetDeposited(MultiAssets([MultiAsset { id: Concrete(MultiLocation { parents: 2, interior: X1(GlobalConsensus(Kusama)) }), fun: Fungible(15) }])),
    ClearOrigin,
    // account from remote_destination reanchored how Statemint will see it
    DepositAsset {
        assets: Wild(All),
        beneficiary: MultiLocation {
                parents: 0,
                interior: X1(AccountId32 { network: Some(Polkadot), id: [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2] }) } }]) 
```

Reply:
@vgeddes 

so, what is `AssetManagerContract`? how many contracts will there be? if only one, that the correct settings could be:

I tried something like this in test for your case:

```
allowed_target_location: MultiLocation::new(
						2,
						X2(
							GlobalConsensus(Ethereum { chain_id: 12345 }),
							Junction::AccountId32 {
								network: None,
                                                                // AssetManagerContract 
								id: AccountId32::new([3; 32]).into()
							}
						)
					),
```

```
transfer_asset_via_bridge(
    // KSM asset from Statemine point of view
    MultiAsset {
        fun: Fungible(1000.into()),
        id: Concrete(MultiLocation { parents: 1, interior: Junctions::Here }),
    },
    // remote destination
    MultiLocation::new(
				2,
				X3(
					GlobalConsensus(Ethereum { chain_id: 12345 }),
                                        // AssetManagerContract  
					Junction::AccountId32 { network: None, id: AccountId32::new([3; 32]).into() },
                                        // some Ethereum Alice account 
					Junction::AccountId32 { network: None, id: AccountId32::new([5; 32]).into() },
				),
			))
)
```

```
Xcm([
    UnpaidExecution { weight_limit: Unlimited, check_origin: None },
    // KSM relay location reanchored how Statemint will see it
    ReserveAssetDeposited(MultiAssets([MultiAsset { id: Concrete(MultiLocation { parents: 2, interior: X1(GlobalConsensus(Kusama)) }), fun: Fungible(15) }])),
    ClearOrigin,
    // account from remote_destination reanchored how Ethereum smart contract or eth-pallet will see it
    DepositAsset {
        assets: Wild(All),
        beneficiary: MultiLocation {
                parents: 0,
                interior: X1(AccountId32 { network: None, id: [5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5] })
    } }])
```

Reply:
Hi @bkontur, sorry about the tardy reply here. I lost track of this thread. Also a lot of time has passed so some details have changed.

To be more clear, for the case of Statemint -> Ethereum, the remote destination location will generally not contained within `allowed_target_location`.

Our XCM executor on Ethereum (the allowed_target_location) is a now single contract (now called "InboundChannel"). It will be represented using
```
/GlobalConsensus(Ethereum)/AccountKey20(InboundChannel)
````

If I control an destination account on Ethereum with a private key, it is represented in XCMv3 notation with:
```
/GlobalConsensus(Ethereum)/AccountKey20(Vincent)
```

So the destination multilocation for the funds is not contained within the `allowed_target_location` multilocation

However, I think recent developments had made this discussion moot. If Statemint is going to support connections to multiple Ethereum bridges on BridgeHub, then this pallet needs some way to choose which bridge to use. Which means practically we will need to "hack" the bridge ID into XCMv3 MultiLocations. Like for example:

```
/GlobalConsensus(Ethereum)/GeneralIndex(Bridge1)/AccountKey20(InboundChannel)
/GlobalConsensus(Ethereum)/GeneralIndex(Bridge1)/AccountKey20(Vincent)
```

So with `allowed_target_location = /GlobalConsensus(Ethereum)/GeneralIndex(Bridge1)`, then your existing code should work fine (I think)

Reply:
@vgeddes 
ok, got your point, I put TODO for this, 
let me finish other stuff, and then I will come back to this,
and we will discuss and set it up "somehow" to satisfy everything.
e.g. there is a ExporterFor which chooses bridge-hub location according to the remote_location, so we have possibilities:
https://github.com/paritytech/cumulus/pull/2013/files#diff-e33e5afc30ebd89df96d4630222efd1c686a38185d23c958d94d5f6fbebdf6feR497-R507